src/main/scala/cognite/spark/v1/RelationshipsRelation.scala

```diff
@@ -0,0 +1,156 @@
+    with WritableRelation {
+  import CdpConnector._
+
+  override def schema: StructType = structType[Relationship]
+
+  override def toRow(a: Relationship): Row = asRow(a)
+
+  override def uniqueId(a: Relationship): String = a.externalId
+
+  override def getStreams(filters: Array[Filter])(
+      client: GenericClient[IO],
+      limit: Option[Int],
+      numPartitions: Int): Seq[Stream[IO, Relationship]] = {
+    val fieldNames =
+      Array(
+        "sourceExternalId",
+        "sourceType",
+        "subtype",
+        "targetExternalId",
+        "targetType",
+        "minStartTime",
+        "maxStartTime",
+        "minEndTime",
+        "maxEndTime",
+        "minActiveAtTime",
+        "maxActiveAtTime",
+        "minConfidence",
+        "maxConfidence",
+        "dataSetId",
+        "labels",
+        "minCreatedTime",
+        "maxCreatedTime",
+        "minLastUpdatedTime",
+        "maxLastUpdatedTime"
+      )
+    val pushdownFilterExpression = toPushdownFilterExpression(filters)
+    val shouldGetAllRows = shouldGetAll(pushdownFilterExpression, fieldNames)
+    val filtersAsMaps = pushdownToParameters(pushdownFilterExpression)
+
+    val relationshipsFilterSeq = if (filtersAsMaps.isEmpty || shouldGetAllRows) {
+      Seq(RelationshipsFilter())
+    } else {
+      filtersAsMaps.distinct.map(relationshipsFilterFromMap)
+    }
+
+    val streamsPerFilter = relationshipsFilterSeq
+      .map { f =>
+        client.relationships.filterPartitions(f, numPartitions, limit)
+      }
+
+    // Merge streams related to each partition to make sure duplicate values are read into
+    // the same RDD partition
+    streamsPerFilter.transpose
+      .map(s => s.reduce(_.merge(_)))
+  }
+
+  def relationshipsFilterFromMap(m: Map[String, String]): RelationshipsFilter =
+    RelationshipsFilter(
+      sourceExternalIds = m.get("sourceExternalIds").map(stringSeqFromWrappedArray),
+      sourceTypes = m.get("sourceTypes").map(stringSeqFromWrappedArray),
+      targetExternalIds = m.get("targetExternalIds").map(stringSeqFromWrappedArray),
+      targetTypes = m.get("targetTypes").map(stringSeqFromWrappedArray),
+      dataSetIds = m.get("dataSetId").map(idsFromWrappedArray(_).map(CogniteInternalId)),
+      startTime = timeRangeFromMinAndMax(m.get("minStartTime"), m.get("maxStartTime")),
+      endTime = timeRangeFromMinAndMax(m.get("minEndTime"), m.get("maxEndTime")),
+      activeAtTime = timeRangeFromMinAndMax(m.get("minActiveAtTime"), m.get("maxActiveAtTime")),
+      labels = getLabelsFilter(m.get("labels")),
+      confidence = confidenceRangeFromMinAndMax(m.get("minConfidence"), m.get("maxConfidence")),
+      createdTime = timeRangeFromMinAndMax(m.get("minCreatedTime"), m.get("maxCreatedTime")),
+      lastUpdatedTime = timeRangeFromMinAndMax(m.get("minLastUpdatedTime"), m.get("maxLastUpdatedTime"))
+    )
+
+  override def insert(rows: Seq[Row]): IO[Unit] = {
+    val relationships = rows.map(fromRow[RelationshipCreate](_))
```

Comment:
Here I am curious how RelationshipCreate would be translated into Row. Type of labels is a class we defined, how does it work? Should I add something to SparkSchemaHelperImpl.fromRow?
```suggestion
final case class RelationshipCreate(
    externalId: String,
    sourceExternalId: String,
    sourceType: String,
    targetExternalId: String,
    targetType: String,
    startTime: Option[Instant] = None,
    endTime: Option[Instant] = None,
    confidence: Option[Double] = None,
    dataSetId: Option[Long] = None,
    labels: Option[Seq[CogniteExternalId]] = None
) extends WithRequiredExternalId```

Reply:
Very good question, and I'm not sure.. the `fromRow` function is based on Scala magic called `macros`.. does it compile?

Reply:
But to answer some of your question - the compiler does a pre-run where it generates the fromRow function _based on the class that we pass it_, so there will exist a particular Scala implementation of the `fromRow[RelationshipCreate]` (so not quite the same as using generics to create abstract functions). This function will be able to take a Spark `Row` object and translate it into a `RelationshipCreate`, as long as that `Row` has the structure that we expect.

Reply:
Yes, it did not complain but I could not try creating anything here since again I could not figure what to do with labels again.